### ISSUE TITLE
`del list[i]` results in incorrect Unbound on subsequent access

### ISSUE THREAD

**Issue by earwig** [microsoft/pyright#6026]
**Describe the bug**
`del list[i]` should behave like `list.pop(i)`, i.e. subsequent accesses to that index should be legal, but pyright treats it like deleting a key from a mapping.

**Code or Screenshots**

```python
L = [1, 2, 3]
del L[0]
reveal_type(L[0])  # should be 'int', but is 'Unbound'
```
<img width="392" alt="Screenshot 2023-09-25 at 11 27 47 PM" src="https://github.com/microsoft/pyright/assets/637235/548617f6-b4c3-4fd6-8803-e9e2d546e0b0">

**VS Code extension or command-line**
Pylance v2023.9.20

**Comment by erictraut** [comment]
I agree this is a false positive error, but I don't think there's anything we can do to address it without eliminating a useful feature: type narrowing for indexed expressions. Removing this feature would result other false positive errors — many more than it would eliminate.

Pyright applies type narrowing to indexed expressions if the subscript is a positive integer literal, like `L[0]`. The assumption is that the type of this expression will remain the same the next time it is used. In practice, there are edge cases (like the one in your example) where this may not be true. This can lead to incorrect assumptions. However, such type narrowing is useful, so this is a tradeoff that pyright makes. 

Consider the following code:

````python
def func(y: int):
    x: list[int | None] = [None, None]
    reveal_type(x[0])  # int | None
    x[0] = y # Type of expression "x[0]" is narrowed to int
    reveal_type(x[0])  # int
    x[0] += 1  # Without type narrowing, this would be flagged as an error

    x.pop(0) # Pyright doesn't know that this invalidates assumptions about narrowed types
    print(x[0]) # None
    reveal_type(x[0])  # int (which is incorrect)
```

The Python type system provides no way for a class to indicate which operations (like `pop` or `__del__`) may invalidate assumptions about indexed operations. Short of hard-coding knowledge of specific methods on specific classes (which is something pyright doesn't do), assumptions need to be made, and they will be wrong some of the time.

As you point out, `del L[0]` is the same as `L.pop(0)`. The latter is more idiomatic in Python, and if you switch to that, you can work around this issue.

I'll think about this a bit more, but I don't currently see any good way to fix this. My inclination is to close it as "won't fix".

**Comment by erictraut** [comment]
After thinking about this more, I found a way to address this issue. I decided to treat any `del` that targets an index expression to "kill" any type narrowing associated with that index. This is a special case because it makes index expressions act slightly different from member access and name expressions for purposes of type narrowing, but I think it's merited in this case.

This change will be included in the next release.

**Comment by earwig** [comment]
Thank you!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.330, which I just published. It will also be included in an upcoming release of pylance.
